### PR TITLE
ROX-34920: track symlink events

### PR DESCRIPTION
**Pull request by Molter73**
## Description

Add a LSM hook for path_symlink, allowing us to handle events that create and modify symlinks in monitored directories. We also change the host scanner methods to use `symlink_metadata` in order to properly capture these types of files. This hook does not have the required inode for the symlink, so generating the event is delegated to `d_instantiate` instead.

While working on delegating to `d_instatiate` some refactoring was done:
* Add `path_read` helpers allowing for reading `bound_path_t` into arbitrary buffers.
* Rewrote most `path_read` methods to use these new helpers.
* Add helpers for getting and deleting entries from the `d_instantiate_ctx` map.
* Cleanup the d_instantiate` and `path_mkdir` hooks with the new helper functions.
* Remove `should_track_mkdir` and replaced it with a call to `is_monitored` passing `NULL` as the first argument.

## Checklist
- [x] Patch has a change log entry **OR** does not need one.
- [x] Investigated and inspected CI test results
- [ ] Updated documentation accordingly

**Automated testing**
  - [ ] Added unit tests
  - [x] Added integration tests
  - [ ] Added regression tests

If any of these don't apply, please comment below.

## Testing Performed

Added integration tests for multiple cases.

## Summary by CodeRabbit

- **New Features**
  - Added end-to-end monitoring for symbolic link creation and updates.
  - Symlink events include link targets, source paths, inode details, and monitoring status.
  - Host scanning follows symlinks and refreshes file mappings when links change.
  - Added symlink activity metrics and telemetry support.

- **Bug Fixes**
  - Improved handling of relative, absolute, Unicode, ignored, mounted, and container-based symlink paths.
  - Scanning now logs and skips individual metadata or pattern errors instead of stopping entirely.

- **Tests**
  - Added comprehensive coverage for symlink creation, overwriting, target resolution, and path edge cases.

### CODE REVIEW

File: fact-ebpf/src/bpf/main.c [hunk 1]
```diff
@@ -601,3 +605,38 @@ int BPF_PROG(trace_move_mount, struct path* from, struct path* to) {
...
+    bpf_printk("Failed to get d_instantiate context entry");
+    goto error;
+  }
+
+  if (path_read_into_append_d_entry(dir, dentry, &symlink_ctx->path, path_hooks_support_bpf_d_path) == NULL) {
+    bpf_printk("Failed to read path");
+    goto error;
+  }
+
+  symlink_ctx->parent_inode = inode_to_key(dir->dentry->d_inode);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply fact-ebpf/src/bpf/main.c:624]: @coderabbit the call to `BPF_CORE_READ` is only needed to keep the verifier happy when accessing non-BTF pointers. Because `dir` is passed into the LSM hook, the compiler automatically adds relocation without needing the explicit call to `BPF_CORE_READ`, this is also true for the `path_mkdir` hook, I'll look into changing it in a follow up PR.

File: fact-ebpf/src/bpf/maps.h [hunk 2]
```diff
@@ -83,12 +83,55 @@ struct {
...
+};
+
 struct {
   __uint(type, BPF_MAP_TYPE_LRU_HASH);
   __type(key, __u64);
-  __type(value, struct mkdir_context_t);
+  __type(value, struct d_instantiate_ctx_t);
   __uint(max_entries, 16384);
-} mkdir_context SEC(".maps");
+} d_instantiate_ctx SEC(".maps");
```
[the remark this answers is not in the document]
Comment (Molter73) [reply fact-ebpf/src/bpf/maps.h:100]: I can't think of a way we could redesign this buffer ATM, I'll look into reducing the number of entries by default and make it configurable in case we need to tweak it in production, but will do so in a follow up PR.

[the remark this answers is not in the document]
Comment (Molter73) [reply fact-ebpf/src/bpf/maps.h:100]: @coderabbit follow up PR is already up: #1453

**Comment by Molter73** [comment]
Looks like the symlink tests are very flaky, will set the PR back as open once I've managed to figure out why and make them more reliable.

File: tests/test_path_symlink.py [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+                host_path=link,
+            ),
+        ]
+    )
+
+
+@pytest.mark.skip(
+    reason='symlinks with absolute paths are broken when '
+    + 'running inside container'
+)
```
[the remark this answers is not in the document]
Comment (Molter73) [reply tests/test_path_symlink.py:207]: This won't be addressed in this PR.

File: fact-ebpf/src/bpf/main.c [hunk 4]
```diff
@@ -601,3 +601,38 @@ int BPF_PROG(trace_move_mount, struct path* from, struct path* to) {
...
+  symlink_ctx->event_type = FILE_ACTIVITY_SYMLINK;
+
+  if (bpf_probe_read_str(symlink_ctx->symlink_target, PATH_MAX, old_name) < 0) {
+    bpf_printk("Failed to read old_name");
+    goto error;
+  }
+
```
Comment (Stringy) [inline fact-ebpf/src/bpf/main.c]: aren't we missing a submit or an inode update here?

**Review by Stringy — APPROVED** [review]
